Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+    withDiagnosticDecorator decorator: DiagnosticDecorator = BasicDiagnosticDecorator()
+  ) -> String {
+    let formatter = Self(contextRange: contextRange, diagnosticDecorator: decorator)
+    return formatter.annotatedSource(
+      inSyntaxTree: tree,
+      withDiagnostics: diagnostics,
+      usingIndentString: indentString,
+      appendingSuffixTexts: suffixTexts,
+      employingSourceLocationConverter: sourceLocationConverter
+    )
+  }
+}
+
+/// Extends `DiagnosticsFormatter` with internal helper types for working with diagnostics.
+extension DiagnosticsFormatter {
+  /// Consolidates different types of diagnostic information, such as messages and notes, under a unified structure.
+  ///
+  /// `FlattenDiagnostic` offers a unified approach to managing and processing various forms of diagnostic information.
+  /// It abstracts away the differences between diagnostic messages and informational notes, allowing for seamless integration
+  /// and handling within the diagnostics formatting process.
+  private enum FlattenDiagnostic {
+    /// Represents a diagnostic message.
+    case diagnostic(Diagnostic)
+
+    /// Represents an informational note from a diagnostic(
+    case note(Note)
+
+    /// Returns the message text associated with the diagnostic entity.
+    var message: String {
+      switch self {
+      case let .diagnostic(diagnostic):
+        return diagnostic.message
+      case let .note(note):
+        return note.message
+      }
+    }
+
+    /// Specifies the severity level of the diagnostic entity.
+    var severity: DiagnosticSeverity {
+      switch self {
+      case let .diagnostic(diagnostic):
+        return diagnostic.diagMessage.severity
+      case .note:
+        return .note
+      }
+    }
+
+    /// Location of diagnostic entity in the source code
+    func location(using converter: SourceLocationConverter) -> SourceLocation {
+      switch self {
+      case let .diagnostic(diagnostic):
+        return diagnostic.location(converter: converter)
+      case let .note(note):
+        return note.location(converter: converter)
+      }
+    }
+  }
+
+  /// A representation of a single source line, potentially enriched with diagnostics or annotations.
+  ///
+  /// `AnnotatedSourceLine` encapsulates a line of source code alongside any associated diagnostics (errors, notes, etc.)
+  /// or annotations for enhanced code readability and insight. The diagnostics are represented by the `FlattenDiagnostic`
+  /// enum, which generalizes the handling of diagnostic messages and informational notes, providing a unified
+  /// approach to represent and process various types of diagnostic information.
+  ///
+  /// - Note: This struct is designed for internal use by the `DiagnosticsFormatter` to organize and manage
+  ///         source lines with their related diagnostics and annotations, facilitating a clearer understanding
+  ///         of code issues and contextual notes.
```

Comment:
Nit

```suggestion
  /// - Note: This struct is designed for internal use by the `DiagnosticsFormatter` to organize and manage
  ///  source lines with their related diagnostics and annotations, facilitating a clearer understanding
  ///  of code issues and contextual notes.
```

Reply:
New lines should be aligned to `Note`? 
Like so:
<img width="954" alt="Screenshot 2024-02-21 at 2 42 20 PM" src="https://github.com/apple/swift-syntax/assets/44930823/baa06d4f-0138-41c8-ad3c-95a2b1d52ead">

And for `Returns:` like so:
<img width="1240" alt="Screenshot 2024-02-21 at 2 44 17 PM" src="https://github.com/apple/swift-syntax/assets/44930823/d71aed7b-84cf-4584-8b81-399a8d3c76b0">

?

Reply:
Yes, that’s how I align them because that’s how I would align any markdown list.